cpp/src_prims/cache/cache.h

```diff
@@ -0,0 +1,314 @@
+*
+* The index management can be considered as a hash map<int, int>, where the int
+* keys are the original vector indices that we want to store, and the values are
+* the cache location of these vectors. The indices are hashed into a bucket
+* whose size equals the associativity. These are the cache sets. If a cache
+* set is full, then new indices are stored by replacing the oldest entries.
+*
+* Using this index mapping we implement methods to store and retrive data from
+* the cache buffer, where a unit of data that we are storing is math_t[n_vec].
+* For example in SVM we store full columns of the kernel matrix at each cache
+* entry.
+*
+* Note: we should have a look if the index management could be simplified using
+* concurrent_unordered_map.cuh from cudf.
+*/
+template <typename math_t, int associativity = 32>
+class Cache {
+ public:
+  /**
+   * Construct a Cache object
+   *
+   * @tparam math_t type of elements to be cached
+   * @tparam associativity number of vectors in a cache set
+   *
+   * @param handle
+   * @param stream
+   * @param n_vec number of elements in a single vector that is stored in a
+   *   cache entry
+   * @param cache_size in MiB
+   */
+  Cache(std::shared_ptr<deviceAllocator> allocator, cudaStream_t stream,
+        int n_vec, float cache_size = 200)
+    : allocator(allocator),
+      n_vec(n_vec),
+      cache_size(cache_size),
+      cache(allocator, stream),
+      vec_idx(allocator, stream),
+      cache_time(allocator, stream),
+      is_cached(allocator, stream),
+      ws_tmp(allocator, stream),
+      idx_tmp(allocator, stream),
+      d_num_selected_out(allocator, stream, 1),
+      d_temp_storage(allocator, stream) {
+    ASSERT(n_vec > 0, "Parameter n_vec: shall be larger than zero");
+    ASSERT(associativity > 0, "Associativity shall be larger than zero");
+    ASSERT(cache_size >= 0, "Cache size should not be negative");
+
+    // Calculate how many vectors would fit the cache
+    int n_cache_vecs = (cache_size * 1024 * 1024) / (sizeof(math_t) * n_vec);
+
+    // The available memory shall be enough for at least one cache set
+    if (n_cache_vecs >= associativity) {
+      n_cache_sets = n_cache_vecs / associativity;
+      n_cache_vecs = n_cache_sets * associativity;
+      cache.resize(n_cache_vecs * n_vec, stream);
+      vec_idx.resize(n_cache_vecs, stream);
+      cache_time.resize(n_cache_vecs, stream);
+      CUDA_CHECK(cudaMemsetAsync(vec_idx.data(), 0,
+                                 vec_idx.size() * sizeof(int), stream));
+      CUDA_CHECK(cudaMemsetAsync(cache_time.data(), 0,
+                                 cache_time.size() * sizeof(int), stream));
+    } else {
+      if (cache_size > 0) {
+        std::cout << "Warning: not enough memory to cache a single set of "
+                     "rows, not using cache\n";
+      }
+      n_cache_sets = 0;
+      cache_size = 0;
+    }
+    //std::cout << "Creating cache with size " << cache_size << " MiB, to store " <<
```

Comment:
Thanks @cjnolet for the comments! Ideally this message would be logged with INFO level. I could remove it, but I would keep it as a reminder that I wanted to log something here. This issue comes up several times in PR #912 too. What is the best practice until we decide about logging #100?

Reply:
The convention in sklearn is to pass a boolean `verbose` argument down to algorithms to enable logging like this. This could further be passed down to the prims.

Will this strategy work for you until we have better logging in place?

Reply:
That should work, I'll implement it tomorrow morning.